### PR TITLE
HDDS-11543. Track OzoneClient object leaks via LeakDetector framework.

### PR DESCRIPTION
**Pull request by sadanand48**
## What changes were proposed in this pull request?

- Track OzoneClient object leaks via LeakDetector framework.
- Fix tests with leaks
- Modify job summary to communicate that the job failed due to memory leak.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-11543

## How was this patch tested?
Unit test added

green CI run : https://github.com/sadanand48/hadoop-ozone/actions/runs/11236369336
CI run with summary for failed job : https://github.com/sadanand48/hadoop-ozone/actions/runs/11234668146/job/31231904269

### CODE REVIEW

**Review by adoroszlai — COMMENTED** [review]
Thanks @sadanand48 for the patch, good idea and mostly looks good.

1. Looks like some tests have leaks, those need to be fixed.

2. Log level for `OzoneClientFactory` should be set to DEBUG to 
get stack trace of leaks, like:

https://github.com/apache/ozone/blob/b3afaec5c1fbb997f1a0bdd563975d2cfeead61f/hadoop-ozone/integration-test/src/test/resources/log4j.properties#L22-L23

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/OzoneRpcClientTests.java [hunk 1]
```diff
@@ -3140,6 +3149,34 @@ void testMultipartUploadOverride(ReplicationConfig replication)
...
+    Arrays.fill(data, (byte) 1);
+    try (OzoneOutputStream out = bucket.createKey(keyName, 10,
+        ReplicationConfig.fromTypeAndFactor(RATIS, ONE), new HashMap<>())) {
+      out.write(data);
+    }
+    client = null;
+    System.gc();
+    GenericTestUtils.waitFor(() -> ozoneClientFactoryLogCapturer.getOutput()
+        .contains("is not closed correctly"), 100, 2000);
+    HddsUtils.setIgnoreReportingLeak(false);
```
Comment (adoroszlai) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/OzoneRpcClientTests.java:3178]: Calling `ozoneClientFactoryLogCapturer.clearOutput()` would also avoid failure in post-processing due to this "leak".  Wouldn't that be better than adding a flag and special output just for the test?

Also, probably should be called in `finally` block.

File: hadoop-ozone/dev-support/checks/_mvn_unit_report.sh [hunk 2]
```diff
@@ -37,7 +37,7 @@ find "." -not -path '*/iteration*' -name 'TEST*.xml' -print0 \
 if [[ "${CHECK:-unit}" == "integration" ]]; then
   find hadoop-ozone/integration-test -not -path '*/iteration*' -name '*-output.txt' -print0 \
       | xargs -n1 -0 "grep" -l -E "not closed properly|was not shutdown properly" \
-      | awk -F/ '{sub("-output.txt",""); print $NF}' \
+      | awk -F/ '{sub("-output.txt",""); print $NF ": Memory Leak detected in test, Failing."}' \
```
Comment (adoroszlai) [inline hadoop-ozone/dev-support/checks/_mvn_unit_report.sh:40]: Let's not add this.  Output is expected to contain test class names, extra description will break processing below.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/AbstractOzoneFileSystemTest.java [hunk 3]
```diff
@@ -2230,7 +2236,17 @@ void testFileSystemWithObjectStoreLayout() throws IOException {
...
+      IllegalArgumentException e =
+          assertThrows(IllegalArgumentException.class, () -> {
+            FileSystem fileSystem = null;
+            try {
+              fileSystem = FileSystem.get(config);
+            } finally {
+              if (fileSystem != null) {
+                fileSystem.close();
+              }
+            }
```
Comment (adoroszlai) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/AbstractOzoneFileSystemTest.java:2248]: Let's add a utility method (e.g. in `GenericTestUtils`) to close the object if expected exception is not thrown.

```java
  public static <T extends Throwable> T assertThrows(
      Class<T> expectedType,
      Callable<? extends AutoCloseable> func) {
    return Assertions.assertThrows(expectedType, () -> {
      final AutoCloseable closeable = func.call();
      try {
        if (closeable != null) {
          closeable.close();
        }
      } catch (Exception ignored) {
      }
    });
  }
```

This let's us keep existing assertion statements (except we need to qualify the call with `GenericTestUtils`:

```java
IllegalArgumentException e = GenericTestUtils.assertThrows(IllegalArgumentException.class,
    () -> FileSystem.get(config));
```

We can also reuse it e.g. for assertions about I/O:

https://github.com/apache/ozone/blob/d1e6900b957348ce50bc13cc5e30df3e1b7605fe/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/AbstractOzoneFileSystemTest.java#L412-L414

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/OzoneRpcClientTests.java [hunk 4]
```diff
@@ -260,9 +265,12 @@ static void startCluster(OzoneConfiguration conf, MiniOzoneCluster.Builder build
    * Close OzoneClient and shutdown MiniOzoneCluster.
    */
   static void shutdownCluster() throws IOException {
-    if (ozClient != null) {
-      ozClient.close();
+    for (OzoneClient ozoneClient : ozoneClients) {
+      if (ozoneClient != null) {
+        ozoneClient.close();
+      }
     }
```
Comment (adoroszlai) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/OzoneRpcClientTests.java:272]:
```suggestion
    org.apache.hadoop.hdds.utils.IOUtils.closeQuietly(ozoneClients);
```

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/shell/TestOzoneShellHA.java [hunk 5]
```diff
@@ -1149,8 +1148,6 @@ public void testListBucket() throws Exception {
             getClientConfForOFS(hostPrefix, cluster.getConf());
     int pageSize = 20;
     clientConf.setInt(OZONE_FS_LISTING_PAGE_SIZE, pageSize);
-    URI uri = FileSystem.getDefaultUri(clientConf);
-    clientConf.setBoolean(String.format("fs.%s.impl.disable.cache", uri.getScheme()), true);
```
Comment (adoroszlai) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/shell/TestOzoneShellHA.java:1153]: Is this related?

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneClientAdapterImpl.java [hunk 6]
```diff
@@ -193,18 +193,24 @@ public BasicOzoneClientAdapterImpl(String omHost, int omPort,
...
+      this.bucket = volume.getBucket(bucketStr);
+      bucketReplicationConfig = this.bucket.getReplicationConfig();
+      nextReplicationConfigRefreshTime = clock.millis() + bucketRepConfigRefreshPeriodMS;
+
+      // resolve the bucket layout in case of Link Bucket
+      BucketLayout resolvedBucketLayout =
+          OzoneClientUtils.resolveLinkBucketLayout(bucket, objectStore, new HashSet<>());
+
+      OzoneFSUtils.validateBucketLayout(bucket.getName(), resolvedBucketLayout);
+    } catch (IOException | IllegalArgumentException exception) {
```
Comment (adoroszlai) [inline hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicOzoneClientAdapterImpl.java:207]: Shouldn't we catch all `RuntimeException`?

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java [hunk 7]
```diff
@@ -221,7 +221,15 @@ public BasicRootedOzoneClientAdapterImpl(String omHost, int omPort,
           OzoneConfigKeys.HDDS_CONTAINER_IPC_PORT_DEFAULT);
 
       // Fetches the bucket layout to be used by OFS.
-      initDefaultFsBucketLayout(conf);
+      try {
+        initDefaultFsBucketLayout(conf);
+      } catch (OMException ome) {
```
Comment (adoroszlai) [inline hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java:226]: Shouldn't we catch `RuntimeException`, too?

**Review by adoroszlai — COMMENTED** [review]
Thanks @sadanand48 for updating the patch.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/shell/TestOzoneShellHA.java (hunk 5 above)
Comment (sadanand48) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/shell/TestOzoneShellHA.java:1153]: yes, there is a leak originating from hadoop code which we have no control from Ozone however unsetting this config fixes this (probably because it uses the cached instance which is properly released in the test)

https://github.com/apache/hadoop/blob/dc56fc385a400b648386e9b7aa8dcb899197bfc7/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/shell/PathData.java#L347

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/OzoneRpcClientTests.java (hunk 1 above)
Comment (adoroszlai) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/OzoneRpcClientTests.java:3178]: You are right.

`TestRocksObjectLeakDetector` is disabled in CI for the same reason:

https://github.com/apache/ozone/blob/e00f7ae960ef2c5a5efc8a439e38f5931d8be383/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/hdds/utils/db/managed/TestRocksObjectLeakDetector.java#L37-L44

We can keep the flag + test-specific message, or disable this test case, too.

Comment (sadanand48) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/client/rpc/OzoneRpcClientTests.java]: Disabled it, as I agree too that making change in production code for testing might not be a good idea. We already have TestLeakDetector to test the generic functionality.

File: hadoop-ozone/dev-support/checks/_mvn_unit_report.sh (hunk 2 above)
Comment (sadanand48) [reply hadoop-ozone/dev-support/checks/_mvn_unit_report.sh]: Can we append a message to end of the TestName-output.txt instead of summary.txt  . I am insisting on this because it is hard to understand why the CI failed in such situations

Comment (adoroszlai) [reply hadoop-ozone/dev-support/checks/_mvn_unit_report.sh]: I'll check: HDDS-11551.
